### ISSUE TITLE
Derivations already built are not read-only - potential corruption of store

### ISSUE THREAD

**Issue by arnoldfarkas** [NixOS/nixpkgs#95207]
**Describe the bug**
After a package is built, another package can accidentally modify it (which happened for me, see details below), resulting in a corrupted /nix/store (hash mismatch).

Background info:
The infrastucture we use is set up in a way that we build Nix expressions, and then instantiate and symlink the outputs (so neither nix-env nor nix-shell is used).  We also use GUI applications, where the .desktop files are read, however those use relative path for executables and icons, and in order to work, we need a full path for both. 
There are packages, which use `makeDesktopItem` to generate  .desktop file(s). Since `makeDesktopItem` is making a derivation, it can only set up relative paths (it cannot depend back to the access the `$out` of the actual package which would cause a cyclic dependency). In order to inject the full path to the generated .desktop item, we have modified some packages to `substituteInPlace` in the last build step (to put `$out`/bin in front of executables, and some path starting with `$out/..` for the icons). 
There are packages (`vscode` and `sublime`, for instance, I did not check all of them), which just symlink the .dekstop file from the `makeDesktopItem` output into `$out/share/applications/<name>.desktop`.  Since the `makeDesktopItem` output is not read-only after it has been built, the last build step to add the full path is modifying **another store path**, in other words, corrupting the already built package. This will cause a hash mismatch error when verifying the content (and also means that subsequent builds will further and further modify the .desktop item, since as the `makeDesktopItem` contents are unchanged, it will not get re-built, so the already modified content will be the base of the update.

Is it intentional that the package that is already built can be modified externally, is it an issue for the output of `makeDesktopItem` only, is it normally forbidden just there is something with out setup that allows such modification, or is unintended/unwanted?

**To Reproduce**
Shortened example, I hope I do not make a typo in here:
1. create an overlay (and set it up, or just modify the code in-place):
```
self: super:
{
  sublime = super.sublime.overrideAttrs (o: {
    buildCommand = o.buildCommand + ''
      substituteInPlace $out/share/applications/sublime2.desktop --replace "Exec=sublime2 %F" "Exec=$out/bin/sublime2 %F"
    '';
  });
}
```
2. Build the package
```
$ NIXPKGS_ALLOW_UNFREE=1 nix-build '<nixpkgs>' -A sublime 
```
3. Verify nix-store
```
$ nix-store --verify --check-contents
```

**Expected behavior**
It should not be possible to modify another derivation during build, since hidden corruption can occur - unless this feature is intentional.

**Notify maintainers**

**Metadata**
(cannot copy to clipboard from Firm infrastructure, it is not a sandbox env, and Nix-env is 2.3.1)

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
# a list of nixos modules affected by the problem
module:
```

**Comment by veprbl** [comment]
Duplicate of #31547
